sql/src/main/java/io/crate/execution/engine/aggregation/impl/CollectSetAggregation.java

```diff
@@ -100,19 +114,158 @@ public DataType partialType() {
+     *
+     * eg. for a window definition of CURRENT ROW -> UNBOUNDED FOLLOWING for {1, 2, 2, 2, 3 } the window frames and
+     * corresponding collect_set outputs are:
+     *  {1, 2, 2, 2, 3}  - [1, 2, 3]
+     *  {2, 2, 2, 3}     - [2, 3]
+     *  {2, 2, 3}        - [2, 3]
+     *  {2, 3}           - [2, 3]
+     *  {3}              - [3]
+     */
+    private static class RemovableCumulativeCollectSet extends AggregationFunction<Map<Object, Object>, Object[]> {
+
+        private static final transient int[] ONE_OCCURENCE = new int[]{1};
+
+        private final SizeEstimator<Object> innerTypeEstimator;
+
+        private final FunctionInfo info;
+        private final DataType partialReturnType;
+
+        RemovableCumulativeCollectSet(FunctionInfo info) {
+            this.innerTypeEstimator = SizeEstimatorFactory.create(((ArrayType) info.returnType()).innerType());
+            this.info = info;
+            this.partialReturnType = UncheckedObjectType.INSTANCE;
+        }
+
+        @Nullable
+        @Override
+        public Map<Object, Object> newState(RamAccountingContext ramAccountingContext,
+                                            Version indexVersionCreated,
+                                            BigArrays bigArrays) {
+            ramAccountingContext.addBytes(RamAccountingContext.roundUp(64L)); // overhead for HashMap: 32 * 0 + 16 * 4 bytes
+            return new HashMap<>();
+        }
+
+        @Override
+        public Map<Object, Object> iterate(RamAccountingContext ramAccountingContext,
+                                           Map<Object, Object> state,
+                                           Input... args) throws CircuitBreakingException {
+            Object value = args[0].value();
+            if (value == null) {
+                return state;
+            }
+            upsertOccurrenceForValue(state, value, 1, ramAccountingContext, innerTypeEstimator);
+            return state;
+        }
+
+        private static void upsertOccurrenceForValue(final Map<Object, Object> state,
+                                                     final Object value,
+                                                     final int occurrenceIncrement,
+                                                     final RamAccountingContext ramAccountingContext,
+                                                     final SizeEstimator<Object> innerTypeEstimator) {
+            state.compute(value, (k, v) -> {
+                if (v == null) {
+                    ramAccountingContext.addBytes(
+                        // values size + 32 bytes for entry, 4 bytes for increased capacity, 4 bytes for the count integer
+                        RamAccountingContext.roundUp(innerTypeEstimator.estimateSize(value) + 40L)
+                    );
+                    if (occurrenceIncrement == 1) {
+                        return ONE_OCCURENCE;
```

Comment:
Is this `ONE_OCCURENCE` optimization really worth it?

Reply:
I can imagine the allocations for a window that spans one day's worth of data could get wild if the cardinality of the values is high.

Reply:
Did you run some benchmarks to back this up? I think optimizations that complicate the code like this should've some verification that they are actual optimizations.

Reply:
I didn't benchmark as I wasn't looking to make it faster in any way but to prevent high memory requirements (and yeah, depending on the GC used, configuration etc avoiding promotions). Do you think it makes sense to have one `int[]{1}` instance for every distinct result in a window?

Reply:
also accounting ram on `ONE_OCCURENCE` https://github.com/crate/crate/pull/8789/files#diff-ed6a9ab02f736d2fea0f6baec6f20588R198 isn't correct

Reply:
yes, you're right @seut (still need to account for the value we store and the map capacity but not the for int we store) Pushed a fixup where we only add the array container and int overhead when we create a new one